book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -641,6 +679,18 @@ Zcash structures are encoded using `ZcashSerialize`/`ZcashDeserialize`.
   block.  This would more traditionally be a `(hash, index)` pair, but because
   we store blocks by height, storing the height saves one level of indirection.
 
+- Each incremental tree consists of nodes for a small number of peaks.
+  Peaks are written once, then deleted when they are no longer required.
+  New incremental tree nodes can be added each time the finalized tip changes,
+  and unused nodes can be deleted.
+  We only keep the nodes needed for the incremental tree for the finalized tip.
```

Comment:
I think that both the current implementation in https://github.com/ZcashFoundation/zebra/pull/2407 and `incrementalmerkletree` don't provide this kind of granularity and will require being serialized in its whole. (I can be wrong particularly about `incrementalmerkletree` because it has a lot of additional stuff that I don't understand yet)

Reply:
`incrementalmerkletree` tracks the minimum frontier of nodes and leaves required to continue appending leaves and producing the same root as a full balanced merkle tree, but serializes that frontier all down. The implementation I proposed derived from `librustzcash` tracks and serializes down the full tree (internal nodes and root etc). I originally wanted to do something like the frontier-only method but didn't grok it with sufficient confidence in time, and wanted yo be as consistent and understandable as possible first

Reply:
Let's add a TODO here, and fix it up in the same PR that merges the final code.
```suggestion
- Each incremental tree consists of nodes for a small number of peaks.
  Peaks are written once, then deleted when they are no longer required.
  New incremental tree nodes can be added each time the finalized tip changes,
  and unused nodes can be deleted.
  We only keep the nodes needed for the incremental tree for the finalized tip.
  TODO: update this description based on the incremental merkle tree code
```